tempodb/backend/raw.go

```diff
@@ -40,6 +49,8 @@ type RawWriter interface {
 type RawReader interface {
 	// List returns all objects one level beneath the provided keypath
 	List(ctx context.Context, keypath KeyPath) ([]string, error)
+	// ListBlocks returns all blockIDs and compactedBlockIDs for a tenant.
+	ListBlocks(ctx context.Context, keypath KeyPath) (blockIDs []uuid.UUID, compactedBlockIDs []uuid.UUID, err error)
```

Comment:
`ListBlocks` is not really _raw_, it's leaking backend implementation details. It could be made more generic with a callback if necessary to do the dir depth check,

```go
		parts := strings.Split(tenantFilePath, "/")
		// i.e: <tenantID/<blockID>/meta
		if len(parts) != 3 {
			return nil
		}
```

and merge the `switch` statement to `backend.Reader`

```go
		switch parts[2] {
		case backend.MetaName:
			metas = append(metas, id)
		case backend.CompactedMetaName:
			compactedMetas = append(compactedMetas, id)
		}
```

Reply:
Not a super strong opinion, just think it'll be a cleaner interface.

Reply:
I thought so too since this was the last approach I'd taken, but it came up in review that folks didn't like the concurrency living one layer above the raw.  See this commit for what was removed.  https://github.com/grafana/tempo/pull/2652/commits/c1da3a6416a36eb14f6225164fbfcfd9b0315d05

Not all backeneds support starting a list call in the middle, so it was necessary to know what backend we were working with.  This was part of the objection to this approach also.  This discovery was removed in this commit.
https://github.com/grafana/tempo/pull/2652/commits/d9145775d28bc9a7978980aecbeb043a13a432af

Do you have ideas about how we might approach this differntly?  I struggled to find a generic solution that could:
1. parallelize the calls
2. avoid mentioning anything about UUID or Blocks in the RawReaders
3. make something that was generic enough to be used in the UUID case (0-f searching) and in a general find case (0-9a-zA-Z) where the number of workers would take an even amount of the load depending on the use case.

The idea of providing "hits" to the backend about what to search within came up, but this felt like we were just dancing around the idea of using the word blocks in the backend.

What do you think?